app/src/lib/fuzzy-find.ts

```diff
@@ -0,0 +1,49 @@
+import * as fuzzAldrin from 'fuzzaldrin-plus'
+
+import { compareDescending } from './compare'
+
+export const options: fuzzAldrin.IFilterOptions = {
+  allowErrors: true,
+  isPath: true,
+  pathSeparator: '-',
+}
+
+export function score(
+  str: string,
+  query: string,
+  maxScore: number = fuzzAldrin.score(query, query, undefined, options)
+) {
+  return fuzzAldrin.score(str, query, undefined, options) / maxScore
+}
+
+export interface IMatch<T> {
+  /** `0 <= score <= 1` */
+  score: number
+  item: T
+  matches: ReadonlyArray<number>
+}
+
+export function match<T, _K extends keyof T>(
```

Comment:
Any reason the generic type is prefixed with an underscore (`_K`)?

Reply:
It’s private, but there’s no way to specify that. I tried `getKey: keyof T | ((item: T) => string)`, but TS didn’t like it. It [thinks](http://www.typescriptlang.org/play/index.html#src=const%20fuzzAldrin%20%3D%20%7B%0A%20%20score%3A%20(...args)%20%3D%3E%203%2C%0A%20%20match%3A%20(...args)%20%3D%3E%20%5B%5D%0A%7D%0A%0Afunction%20compareDescending(left%2C%20right)%20%7B%20return%200%20%7D%0A%0Aexport%20const%20options%20%3D%20%7B%0A%20%20allowErrors%3A%20true%2C%0A%20%20isPath%3A%20true%2C%0A%20%20pathSeparator%3A%20'-'%2C%0A%7D%0A%0Aexport%20function%20score(%0A%20%20str%3A%20string%2C%0A%20%20query%3A%20string%2C%0A%20%20maxScore%3A%20number%0A)%20%7B%0A%20%20return%20fuzzAldrin.score(str%2C%20query%2C%20undefined%2C%20options)%20%2F%20maxScore%0A%7D%0A%0Aexport%20interface%20IMatch%3CT%3E%20%7B%0A%20%20%2F**%20%600%20%3C%3D%20score%20%3C%3D%201%60%20*%2F%0A%20%20score%3A%20number%0A%20%20item%3A%20T%0A%20%20matches%3A%20ReadonlyArray%3Cnumber%3E%0A%7D%0A%0Aexport%20function%20match%3CT%3E(%0A%20%20query%3A%20string%2C%0A%20%20items%3A%20ReadonlyArray%3CT%3E%2C%0A%20%20getKey%3A%20(keyof%20T)%20%7C%20((item%3A%20T)%20%3D%3E%20string)%0A)%3A%20ReadonlyArray%3CIMatch%3CT%3E%3E%20%7B%0A%20%20%2F%2F%20matching%20%60query%60%20against%20itself%20is%20a%20perfect%20match.%0A%20%20const%20maxScore%20%3D%20score(query%2C%20query%2C%201)%0A%20%20return%20items%0A%20%20%20%20.map((item)%3A%20IMatch%3CT%3E%20%3D%3E%20%7B%0A%20%20%20%20%20%20let%20key%3A%20string%0A%20%20%20%20%20%20if%20(typeof%20getKey%20%3D%3D%3D%20'function')%20%7B%0A%20%20%20%20%20%20%20%20key%20%3D%20getKey(item)%0A%20%20%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%20%20%20%20key%20%3D%20String(item%5BgetKey%5D)%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20return%20%7B%0A%20%20%20%20%20%20%20%20score%3A%20score(key%2C%20query%2C%20maxScore)%2C%0A%20%20%20%20%20%20%20%20item%2C%0A%20%20%20%20%20%20%20%20matches%3A%20fuzzAldrin.match(key%2C%20query%2C%20undefined%2C%20options)%2C%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D)%0A%20%20%20%20.filter((%7B%20matches%20%7D)%20%3D%3E%20matches.length%20%3E%200)%0A%20%20%20%20.sort((%7B%20score%3A%20left%20%7D%2C%20%7B%20score%3A%20right%20%7D)%20%3D%3E%20compareDescending(left%2C%20right))%0A%7D%0A) that `getKey` can be something other than a function in the `if (typeof getKey === 'function')` block below.

Reply:
I wonder if that's because `typeof` isn't 100% accurate when it comes to types...

Reply:
Should I do `(getKey as (item: T) => string)(item)` instead?

Reply:
I think casting is fine as we (as humans) are pretty sure that `getKey` **is** a function.